File: src/15장/15-2-4.py
from collections import deque

visit=[[0 for col in range(10)] for row in range(10)]
visit2=[[0 for col in range(10)] for row in range(10)]
dx=[0,0,1,-1]
dy=[1,-1,0,0]

def bfs():
  global answer

  for x in range(n):
    for y in range(m):
      visit[x][y]=lab[x][y]
      if(lab[x][y]==2):
        queue.append([x,y])
  
  while(queue):
    x,y=queue.popleft()
    visit[x][y]=1
    for i in range(4):
      if 0<= x+dx[i]<n and 0<=y+dy[i]<m and lab[x+dx[i]][y+dy[i]]==0 and visit[x+dx[i]][y+dy[i]]==0:
        queue.append([x+dx[i],y+dy[i]])
        visit[x+dx[i]][y+dy[i]]=1

  cnt=0
  for x in range(n):
    for y in range(m):
      if lab[x][y]==0 and visit[x][y]==0:
        cnt+=1
  answer=max(answer,cnt)

def back_Traking(select):
    
    if select == 3:
      bfs()
      return
    for x in range(n):
      for y in range(m):
        if not lab[x][y] and not visit2[x][y]:
          visit2[x][y]=1
          lab[x][y] = 1
          back_Traking(select + 1)
          lab[x][y] = 0
          visit2[x][y]=0

queue=deque()
answer = 0
n, m = map(int, input().split())
lab = [list(map(int, input().split())) for _ in range(n)]
back_Traking(0)
print(answer)

File: src/15장/15-2-4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

class Main{
    static int N, M;
    static int[][] lab;
    static int[][] dir = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    static int[][] visit;
    static int answer = 0;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split(" ");
        N = Integer.parseInt(input[0]);
        M = Integer.parseInt(input[1]);
        lab = new int[N][M];
        visit = new int[N][M];
        for (int i = 0; i < N; i++) {
            input = br.readLine().split(" ");
            for (int j = 0; j < M; j++) {
                lab[i][j] = Integer.parseInt(input[j]);
            }
        }
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (lab[i][j] == 0) {
                    copy();
                    visit[i][j] = 1;
                    back_Traking(1);
                    visit[i][j] = 0;
                }
            }
        }
        System.out.println(answer);
    }

    static void copy() {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                visit[i][j] = lab[i][j];
            }
        }
    }
    static void back_Traking(int select) {
        if (select == 3) {
            bfs();
            return;
        }
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if(visit[i][j] == 0) {
                    visit[i][j] = 1;
                    back_Traking(select+1);
                    visit[i][j] = 0;
                }
            }
        }
    }
    static void bfs() {
        Queue<Integer> queueX = new LinkedList<>();
        Queue<Integer> queueY = new LinkedList<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (visit[i][j] == 2) {
                    queueX.offer(j);
                    queueY.offer(i);
                }
            }
        }

        int[][] visit2 = new int[N][M];
        for(int i = 0; i < N; i++) {
            for(int j = 0; j < M; j++) {
                visit2[i][j] = visit[i][j];
            }
        }
        while (!queueX.isEmpty()) {
            int tempX = queueX.poll();
            int tempY = queueY.poll();
            for (int i = 0; i < 4; i++) {
                int destX = tempX + dir[i][1];
                int destY = tempY + dir[i][0];
                if (destX < 0 || destX >= M || destY < 0 || destY >= N) continue;
                if (visit2[destY][destX] != 0) continue;
                visit2[destY][destX] = 2;
                queueX.offer(destX);
                queueY.offer(destY);
            }
        }
        int count = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (visit2[i][j] == 0) count++;
            }
        }
        if (count > answer) answer = count;
    }
}